### PR TITLE
kubectl describe: show annotations

### PR DESCRIPTION
**Pull request by chenchun**
Fixes https://github.com/kubernetes/kubernetes/issues/23380

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35094)

### CODE REVIEW

**Comment by smarterclayton** [comment]
If there are no annotations it may be better to display `<none>` rather than an empty string.

**Comment by smarterclayton** [comment]
@k8s-bot ok to test

**Comment by chenchun** [comment]
Yes, it already prints `<none>` like labels https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/describe.go#L2540 .

**Comment by chenchun** [comment]
ping @smarterclayton

**Comment by kargakis** [comment]
Some annotations are huge. Are we going to trim those?

**Comment by chenchun** [comment]
@kargakis For how huge are you talking about, can you show me some examples ?

**Comment by kargakis** [comment]
SerializedReferences, initContainers as is, and any kind of data that can be stored in an annotation

**Comment by chenchun** [comment]
@kargakis Updated to make describing huge annotations omite chars for more than a fixed length, e.g. 200.

But I think displaying the entire annotations is ok since pod events can also be very long.
Before update
```
[root@kubernetes-master vagrant]# kubectl describe pod hello-1169216084-16ud7
Name:		hello-1169216084-16ud7
Namespace:	default
Node:		kubernetes-node-1/10.245.1.3
Start Time:	Mon, 31 Oct 2016 07:34:25 +0000
Labels:		app=hello
		pod-template-hash=1169216084
Annotations:	kubernetes.io/created-by={"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicaSet","namespace":"default","name":"hello-1169216084","uid":"72f98128-9f3c-11e6-adb8-080027242396","apiVersion":"extensions","resourceVersion":"79865"}}

		kubernetes.io/limit-ranger=LimitRanger plugin set: cpu request for container hello; cpu request for init container install
Status:		Running
IP:		10.246.101.2
Controllers:	ReplicaSet/hello-1169216084
Init Containers:
  install:
    Container ID:	docker://05c4f2e3b1d6784516dea772d83f6461ebad244db8839e5f1044eb5d22e49454
    Image:		busybox
    Image ID:		docker://sha256:e02e811dd08fd49e7f6032625495118e63f597eb150403d02e3238af1df240ba
    Port:		
    Command:
      ip
      ad
    Requests:
      cpu:		100m
    State:		Terminated
      Reason:		Completed
      Exit Code:	0
      Started:		Mon, 31 Oct 2016 07:34:36 +0000
      Finished:		Mon, 31 Oct 2016 07:34:36 +0000
    Ready:		True
    Restart Count:	0
    Volume Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-eo6wu (ro)
    Environment Variables:	<none>
Containers:
  hello:
    Container ID:	docker://af54ab56f4975b0fedf597c1c347a4278bb4c665e8697642c8267ffe1c13781a
    Image:		chenchun/hello
    Image ID:		docker://sha256:e88f7a466dcd2eaca1f8f21d3ca6cb2f34ca2e62b094497e66993e380f73db37
    Port:		80/TCP
    Requests:
      cpu:		100m
      memory:		128Mi
    State:		Running
      Started:		Mon, 31 Oct 2016 07:34:43 +0000
    Ready:		True
    Restart Count:	0
    Volume Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-eo6wu (ro)
    Environment Variables:	<none>
Conditions:
  Type		Status
  Initialized 	True 
  Ready 	True 
  PodScheduled 	True 
Volumes:
  default-token-eo6wu:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	default-token-eo6wu
QoS Class:	Burstable
Tolerations:	<none>
Events:
  FirstSeen	LastSeen	Count	From				SubobjectPath			Type		Reason		Message
  ---------	--------	-----	----				-------------			--------	------		-------
  9m		9m		1	{default-scheduler }						Normal		Scheduled	Successfully assigned hello-1169216084-16ud7 to kubernetes-node-1
  9m		9m		1	{kubelet kubernetes-node-1}	spec.initContainers{install}	Normal		Pulling		pulling image "busybox"
  8m		8m		1	{kubelet kubernetes-node-1}	spec.initContainers{install}	Normal		Pulled		Successfully pulled image "busybox"
  8m		8m		1	{kubelet kubernetes-node-1}	spec.initContainers{install}	Normal		Created		Created container with docker id 05c4f2e3b1d6; Security:[seccomp=unconfined]
  8m		8m		1	{kubelet kubernetes-node-1}	spec.initContainers{install}	Normal		Started		Started container with docker id 05c4f2e3b1d6
  8m		8m		1	{kubelet kubernetes-node-1}	spec.containers{hello}		Normal		Pulling		pulling image "chenchun/hello"
  8m		8m		1	{kubelet kubernetes-node-1}	spec.containers{hello}		Normal		Pulled		Successfully pulled image "chenchun/hello"
  8m		8m		1	{kubelet kubernetes-node-1}	spec.containers{hello}		Normal		Created		Created container with docker id af54ab56f497; Security:[seccomp=unconfined]
  8m		8m		1	{kubelet kubernetes-node-1}	spec.containers{hello}		Normal		Started		Started container with docker id af54ab56f497
```

After update
```
[root@kubernetes-master vagrant]# ./kubectl describe pod hello-1169216084-16ud7
Name:		hello-1169216084-16ud7
Namespace:	default
Node:		kubernetes-node-1/10.245.1.3
Start Time:	Mon, 31 Oct 2016 07:34:25 +0000
Labels:		app=hello
		pod-template-hash=1169216084
Annotations:	kubernetes.io/created-by={"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicaSet","namespace":"default","name":"hello-1169216084","uid":"72f98128-9f3c-11e6-adb8-080027242396","...
		kubernetes.io/limit-ranger=LimitRanger plugin set: cpu request for container hello; cpu request for init container install
Status:		Running
IP:		10.246.101.2
Controllers:	ReplicaSet/hello-1169216084
Init Containers:
  install:
    Container ID:	docker://05c4f2e3b1d6784516dea772d83f6461ebad244db8839e5f1044eb5d22e49454
    Image:		busybox
    Image ID:		docker://sha256:e02e811dd08fd49e7f6032625495118e63f597eb150403d02e3238af1df240ba
    Port:		
    Command:
      ip
      ad
    Requests:
      cpu:		100m
    State:		Terminated
      Reason:		Completed
      Exit Code:	0
      Started:		Mon, 31 Oct 2016 07:34:36 +0000
      Finished:		Mon, 31 Oct 2016 07:34:36 +0000
    Ready:		True
    Restart Count:	0
    Volume Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-eo6wu (ro)
    Environment Variables:	<none>
Containers:
  hello:
    Container ID:	docker://af54ab56f4975b0fedf597c1c347a4278bb4c665e8697642c8267ffe1c13781a
    Image:		chenchun/hello
    Image ID:		docker://sha256:e88f7a466dcd2eaca1f8f21d3ca6cb2f34ca2e62b094497e66993e380f73db37
    Port:		80/TCP
    Requests:
      cpu:		100m
      memory:		128Mi
    State:		Running
      Started:		Mon, 31 Oct 2016 07:34:43 +0000
    Ready:		True
    Restart Count:	0
    Volume Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-eo6wu (ro)
    Environment Variables:	<none>
Conditions:
  Type		Status
  Initialized 	True 
  Ready 	True 
  PodScheduled 	True 
Volumes:
  default-token-eo6wu:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	default-token-eo6wu
QoS Class:	Burstable
Tolerations:	<none>
Events:
  FirstSeen	LastSeen	Count	From				SubobjectPath			Type		Reason		Message
  ---------	--------	-----	----				-------------			--------	------		-------
  13m		13m		1	{default-scheduler }						Normal		Scheduled	Successfully assigned hello-1169216084-16ud7 to kubernetes-node-1
  13m		13m		1	{kubelet kubernetes-node-1}	spec.initContainers{install}	Normal		Pulling		pulling image "busybox"
  13m		13m		1	{kubelet kubernetes-node-1}	spec.initContainers{install}	Normal		Pulled		Successfully pulled image "busybox"
  13m		13m		1	{kubelet kubernetes-node-1}	spec.initContainers{install}	Normal		Created		Created container with docker id 05c4f2e3b1d6; Security:[seccomp=unconfined]
  13m		13m		1	{kubelet kubernetes-node-1}	spec.initContainers{install}	Normal		Started		Started container with docker id 05c4f2e3b1d6
  13m		13m		1	{kubelet kubernetes-node-1}	spec.containers{hello}		Normal		Pulling		pulling image "chenchun/hello"
  13m		13m		1	{kubelet kubernetes-node-1}	spec.containers{hello}		Normal		Pulled		Successfully pulled image "chenchun/hello"
  13m		13m		1	{kubelet kubernetes-node-1}	spec.containers{hello}		Normal		Created		Created container with docker id af54ab56f497; Security:[seccomp=unconfined]
  13m		13m		1	{kubelet kubernetes-node-1}	spec.containers{hello}		Normal		Started		Started container with docker id af54ab56f497
```

**Comment by smarterclayton** [comment]
Can you rebase?  Would like to get this in.

**Comment by chenchun** [comment]
@smarterclayton Rebased and tested manually.

File: pkg/kubectl/describe.go [hunk 1]
```diff
@@ -2904,3 +2918,41 @@ func printTolerationsMultilineWithIndent(w *PrefixWriter, initialIndent, title,
...
+	printAnnotationsMultilineWithIndent(w, "", title, "\t", labels)
+}
+
+// printAnnotationsMultilineWithIndent prints multiple annotations with a user-defined alignment.
+// If annotation string is too long, we omit chars more than 200 length.
+func printAnnotationsMultilineWithIndent(w *PrefixWriter, initialIndent, title, innerIndent string, annotations map[string]string) {
+
+	w.Write(LEVEL_0, "%s%s:%s", initialIndent, title, innerIndent)
+
+	if annotations == nil || len(annotations) == 0 {
```
Comment (smarterclayton) [inline pkg/kubectl/describe.go]: nil check is redundant

File: pkg/kubectl/describe.go [hunk 2]
```diff
@@ -2904,3 +2918,41 @@ func printTolerationsMultilineWithIndent(w *PrefixWriter, initialIndent, title,
...
 }
+
+// printAnnotationsMultiline prints multiple annotations with a proper alignment.
+func printAnnotationsMultiline(w *PrefixWriter, title string, labels map[string]string) {
+	printAnnotationsMultilineWithIndent(w, "", title, "\t", labels)
+}
+
+// printAnnotationsMultilineWithIndent prints multiple annotations with a user-defined alignment.
+// If annotation string is too long, we omit chars more than 200 length.
+func printAnnotationsMultilineWithIndent(w *PrefixWriter, initialIndent, title, innerIndent string, annotations map[string]string) {
```
Comment (smarterclayton) [inline pkg/kubectl/describe.go]: Is there abunit test that exercises this?

**Comment by chenchun** [comment]
@k8s-bot cvm gce e2e test this

**Comment by smarterclayton** [comment]
@k8s-bot cvm gce e2e test this

File: pkg/kubectl/describe.go (hunk 2 above)
Comment (chenchun) [inline pkg/kubectl/describe.go]: Done

**Comment by smarterclayton** [comment]
This has been in the queue for weeks, adding to 1.6 milestone.  

/lgtm

**Pull request merged**